Library/Homebrew/formula_installer.rb

```diff
@@ -151,6 +151,28 @@ def check_install_sanity
     recursive_deps = formula.recursive_dependencies
     recursive_formulae = recursive_deps.map(&:to_formula)
 
+    recursive_dependencies = []
+    recursive_formulae.each do |dep|
+      dep_recursive_dependencies = dep.recursive_dependencies.map(&:to_s)
+      if dep_recursive_dependencies.include?(formula.name)
+        recursive_dependencies << "#{formula.full_name} depends on #{dep.full_name}"
+        recursive_dependencies << "#{dep.full_name} depends on #{formula.full_name}"
+      end
+    end
+
+    unless recursive_dependencies.empty?
```

Comment:
@vladshablinsky They are separate checks for two edge cases that previously weren't handled. Thoughts on how to consolidate would be welcome. Going to merge as-is for now just to get the bug fixed but I'm genuinely interested in your reply and will address any thoughts on a follow-up PR.